Inheritance raytracers

boboeyen opened the pull request:
Inheritance of basic ray tracer implemented to avoid implementing the same functions at multiple places.

cg-laser reviewed (changes requested):
Hey Bob, thanks for starting that. It look like the right way to proceed. I haven't gotten through all changes yet, but I already have one comment.  
Then 2Danalytic ray tracing class should not inherit from the base class. This is a special thing. It is a low level interface of actually calculating the solutions (and just 2D). The class that should inherit from the baseclass is the 3D class (this is just named `raytracing` I think).

anelles commented on NuRadioMC/SignalProp/propagation_base_class.py:
```diff
@@ -1,26 +1,31 @@
 Structure of a ray-tracing module. For documentation and development purposes.
 """
 
-solution_types = {1: 'direct',
```
@boboeyen has this been updated? Then you can mark the conversation as resolved.

cg-laser replied:
in general, it would be good if you quickly responded how you addressed the comment before resolving the comment. (but I found it ;-))

anelles commented:
I think it makes a bit more sense, if folks review this, who use the ray tracing more often. To me, it looks good, but that probably doesn't help.

anelles commented:
In any case, the conflicts need to be resolved before merging.

sjoerd-bouma commented on NuRadioMC/SignalProp/radioproparaytracing.py:
```diff
@@ -74,48 +71,34 @@ class describing the index-of-refraction profile
-        if config != None: 
-            self.__cut_viewing_angle = config['speedup']['delta_C_cut']*units.radian
+        if self._config != None: 
+            self._cut_viewing_angle = config['speedup']['delta_C_cut']*units.radian
         else: 
-            self.__cut_viewing_angle = 40*units.degree
+            self._cut_viewing_angle = 40*units.degree
```
I guess now that the analytic ray tracer has a default config, we should implement this in a similar way for the other ray tracers for consistency?

boboeyen replied:
yes indeed, I saw that to and came to the same conclusion! :)

sjoerd-bouma commented:
I looked through this, and it looks fine to me. Are all three ray tracers used in tests at the moment? If not, I'll just run a couple of tests myself and approve this maybe tomorrow.
One question, out of curiosity - what's the reason for changing the internal properties of the ray tracers (e.g self._config, self._logger) from 2 to 1 underscores? E.g. for the logger I don't necessarily see why you need to access it from external modules.

sjoerd-bouma commented:
The radiopropa ray tracer and direct ray tracer were broken (I used the example [here](https://github.com/nu-radio/NuRadioMC/wiki/Manual:-signal-propagation) to test), so I fixed them. I also added the linked example at SignalProp/examples/ray_tracing_modules.py. It tries to trace some rays and plot the results for all available modules. We may want to consider adding it to the test suite.

@boboeyen, if you resolve Christian's comment and answer my previous question (and maybe quickly glance over my commits to make sure I didn't do anything crazy), I'm happy to approve this PR (finally...)

boboeyen commented:
> I looked through this, and it looks fine to me. Are all three ray tracers used in tests at the moment? If not, I'll just run a couple of tests myself and approve this maybe tomorrow. One question, out of curiosity - what's the reason for changing the internal properties of the ray tracers (e.g self._config, self._logger) from 2 to 1 underscores? E.g. for the logger I don't necessarily see why you need to access it from external modules.

I did this because otherwise the logger/config that is initiated by the super().__init__() function can not be accesed by the daughter class. Maybe I am missing some fancy/better way of doing this

boboeyen commented:
> @boboeyen, if you resolve Christian's comment and answer my previous question (and maybe quickly glance over my commits to make sure I didn't do anything crazy), I'm happy to approve this PR (finally...)

Christian's comments were taken care of normally
Question, see above
Your commits look fine to me

sjoerd-bouma reviewed (approved):
Thanks for answering my question - I think it's fine to use _ instead of __ (you could implement the latter with get_ and set_ functions, which already exist for the config, for example), but I don't really have a strong preference for either.
If anybody else wants to comment, feel free, but it looks good to me.

cg-laser commented:
I'll have a look tomorrow

cg-laser reviewed (changes requested):
looks good to me!

@sjoerd-bouma thanks for adding the example that tests all RT modules for functionality. Can you add it to our test suite? Then the PR is good to go.

cg-laser commented:
I think we should leave the member variables as protected ("_").

sjoerd-bouma commented:
> looks good to me!
> 
> @sjoerd-bouma thanks for adding the example that tests all RT modules for functionality. Can you add it to our test suite? Then the PR is good to go.

This is not _quite_ as straightforward as it sounds, as RadioPropa needs to be installed (& compiled) in order for the test to work...

cg-laser commented:
what about adapting the script to check for an installed radio propa, and to just skip it when its not there? 

In general we should also add a RadioPropa test, but maybe to that in a separate PR? @boboeyen ?

sjoerd-bouma commented:
@boboeyen Argh! I am sure that my script ran fine two weeks ago, but now, without any additional changes, I get an error:
```
Traceback (most recent call last):
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioMC/SignalProp/examples/ray_tracing_modules.py", line 27, in <module>
    rays = prop(ice, attenuation_model,
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioMC/SignalProp/radioproparaytracing.py", line 92, in __init__
    self._ice_model = self._medium.get_ice_model_radiopropa()
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioMC/utilities/medium_base.py", line 260, in get_ice_model_radiopropa
    return RadioPropaIceWrapper(self, scalar_field)
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioMC/utilities/medium_base.py", line 338, in __init__
    bottom_reflection = RP.ReflectiveLayer(RP.Plane(RP.Vector3d(*(reflection_pos*(RP.meter/units.meter))),
TypeError: unsupported operand type(s) for *: 'NoneType' and 'float'
```
The issue seems to be that `RP.IceModel_Simple()` returns an object with attribute 'reflection' set to None; when `RadioPropaIceWrapper.__init__` then checks `if hasattr(self.__ice_model_nurario, 'reflection'):`, it finds `True`, but because it is actually None, setting the reflective layer fails. I have tried reinstalling RadioPropa, with no improvement.

I have no clue why it doesn't work now, but used to work 2 weeks ago. Can you please run the script (`NuRadioMC/SignalProp/examples/ray_tracing_modules.py`) and tell me if it works for you, at least?

boboeyen commented:
@sjoerd-bouma I pulled in the last new commits and the example you're referring to is not working for me either. But I think I see what the problem is. 

You use the same ice object and pass it iterative to different propagators. First it is passed to the analytic tracer but if the NuRadio ice object does not have a reflection attribute, the analytic raytracer creates one and sets it to None. However, in the next iteration of the radiopropa tracer, the RadioPropaIceWrapper assumes that if there is a reflection attribute in the ice object it is also set and not None. I can fix this for you

sjoerd-bouma commented:
> @sjoerd-bouma I pulled in the last new commits and the example you're referring to is not working for me either. But I think I see what the problem is.
> 
> You use the same ice object and pass it iterative to different propagators. First it is passed to the analytic tracer but if the NuRadio ice object does not have a reflection attribute, the analytic raytracer creates one and sets it to None. However, in the next iteration of the radiopropa tracer, the RadioPropaIceWrapper assumes that if there is a reflection attribute in the ice object it is also set and not None. I can fix this for you

Thanks! I "fixed" it now by also checking for `.reflection != None` in `medium_base`. If you feel like this is a bad idea, feel free to revert it and implement your own fix.

boboeyen commented:
Okay, then I think thing are ready for merging right? Only @cg-laser has to approve now :)

boboeyen commented:
I just added a final check for the analytic ray tracer so it trows an error if the ice model is not a simple model (as more complex models will come) to give the user some more information on why some stuff is ot working. Also changed the logger of the ray tracers from protected to private so a user is more aware of where an error is produced: mother or daughter class). Hopefully this is fine

boboeyen commented:
I just added a final check for the analytic ray tracer so it trows an error if the ice model is not a simple model (as more complex models will come) to give the user some more information on why some stuff is ot working. Also changed the logger of the ray tracers from protected to private so a user is more aware of where an error is produced: mother or daughter class). Hopefully this is fine

cg-laser reviewed (approved):
looks good, but please wait until the check pass before merge
